packages/editor/src/components/block-settings-menu/index.js

```diff
@@ -148,6 +184,50 @@ export class BlockSettingsMenu extends Component {
 
 export default compose( [
 	withDeprecatedUniqueId,
+	withSelect( ( select, { clientIds, rootClientId } ) => {
+		const {
+			getBlocksByClientId,
+			getBlockIndex,
+			getTemplateLock,
+		} = select( 'core/editor' );
+
+		const blocks = getBlocksByClientId( clientIds );
+
+		const canDuplicate = every( blocks, ( block ) => {
+			return !! block && hasBlockSupport( block.name, 'multiple', true );
+		} );
+
+		return {
+			index: getBlockIndex( last( castArray( clientIds ) ), rootClientId ),
+			isLocked: !! getTemplateLock( rootClientId ),
+			blocks,
+			canDuplicate,
+			shortcuts,
+		};
+	} ),
+	withDispatch( ( dispatch, { clientIds, rootClientId, blocks, index, isLocked, canDuplicate } ) => ( {
+		onDuplicate() {
+			if ( isLocked || ! canDuplicate ) {
+				return;
+			}
+
+			const clonedBlocks = blocks.map( ( block ) => cloneBlock( block ) );
+			dispatch( 'core/editor' ).insertBlocks(
+				clonedBlocks,
+				index + 1,
+				rootClientId
+			);
+			if ( clonedBlocks.length > 1 ) {
+				dispatch( 'core/editor' ).multiSelect(
+					first( clonedBlocks ).clientId,
+					last( clonedBlocks ).clientId
+				);
+			}
+		},
+		onRemove() {
+			dispatch( 'core/editor' ).removeBlocks( clientIds );
```

Comment:
Minor: Using destructuring in our `withSelect` and `withDispatch` callbacks is a nice way to reduce how many times we call `select()` and `dispatch()`. I think it reads a little nicer too—but that's subjective, of course.

```js
withDispatch( ( dispatch, { clientIds, ... } ) => {
	const { insertBlocks, multiSelect, removeBlocks } = dispatch( 'core/editor' );

	return {
		onDuplicate() {
			...
		},
		onRemove() {
			removeBlocks( clientIds );
		},
	};
} )
```